### PR TITLE
Add an option to replace external-lib-deps command

### PR DESCRIPTION
**Pull request by moyodiallo**
This PR is motivate by [dune-opam-lint](https://github.com/ocurrent/opam-dune-lint). The command `dune external-lib-deps` is used by it and this command is removed in dune #4298. Getting the command back is not realistic, I think. Instead of getting all information by doing a static analysis of the build graph. We extract all the information we need during the build, so `dynamically`.

* The first approach is to categorize the Data Base dependencies like `Project_libs` db, `Public_libs` db, `Installed_libs` db. Then, during the resolution catch all libs that searched in `Installed_libs`, potentially there are external libs. Easy to add but you have to propagate the `build_dir` in order to get the dependencies by `build_dir`.
* The second approach, after resolving the lib we could get the necessary needed info from  `Lib_info`. So you don't have to propagate the `build_dir` across many functions, but the issue is you could miss external dependency (like `select` rule). In the case of a dune project not all external deps will be resolved in some particular build.

In this PR the first approach was chosen, why not propagate the `build_dir` and get all external dependencies during the build.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Thanks for explaining the implementation strategy. I still don't under the point of this `dir` argument though. Could you clarify how it helps us deal with conditional dependencies?

**Comment by rgrinberg** [comment]
I should note that using mutation and global variables is also highly suspicious. We've worked very hard to remove all mutable state in our codebase to make it memoization safe. It's very unlikely we're going to undo all that work to reimplement `external-lib-deps`

**Comment by moyodiallo** [comment]
> Thanks for explaining the implementation strategy. I still don't under the point of this `dir` argument though. Could you clarify how it helps us deal with conditional dependencies?

The `dir` argument is to know a dependency belong to which package, because the name of a given package  (`*.opam`) could be in the `dir`, that why we need this information.

**Comment by rgrinberg** [comment]
Where is the association from directory to package is done? I don't see in the code.

For that matter, I don't see why we need to be concerned with packages at all. The check works on libraries only.

**Comment by moyodiallo** [comment]
> I should note that using mutation and global variables is also highly suspicious. We've worked very hard to remove all mutable state in our codebase to make it memoization safe. It's very unlikely we're going to undo all that work to reimplement `external-lib-deps`

True. This is a first shot, I did think about that. Is Memoizing them going to be better in this case ?

**Comment by rgrinberg** [comment]
Memoization is an optimization, so it is something we can think about if our implementation is slow. For now, I'd suggest just trying to extract the information you need by climbing the workspace and building what is needed to extract lib deps (kind of like dune describe for example). There might not be enough information, or the api is insufficient. In that case, we can tweak `Lib.t` (or whatever else) to provide what you need.

**Comment by moyodiallo** [comment]
> Where is the association from directory to package is done? I don't see in the code.
> 
> For that matter, I don't see why we need to be concerned with packages at all. The check works on libraries only.

Sorry I didn't precise, those information are important to `opam-dune-lint`. This is trying to match as much as the result of the removed command `dune external-lib-deps --sexp --unstable-by-dir @install`. `opam-dune-lint` extract the dependencies for each package (`*.opam` files), from the results of `--external-lib-deps` or `dune external-lib-deps`.

**Comment by emillon** [comment]
Before discussing the implementation I think it's important to discuss the design of the features and the expectations we have for it.

- First, the UI. It seems that it's a CLI flag to add to `build` and that will display the missing external libs. How is it intended to be used: just `dune build --external-lib-deps`? We're already displaying an error message when an external library is not found. What is the expected difference with this flag?
- I think I would have expected something a separate command, or a new output in `dune describe`. I understand that this is less precise: we can't find dependencies of a single target and we can't see behind `(select)` but is it important to include external libraries that are not going to be selected by the build? But we can probably filter by package by considering only the relevant rules, which seems enough to start.
- I don't understand the `dir` argument either. what does this enable? can we start without this? it seems to me that if we first try to collect all external lib dependencies referred to by `@install` or `@runtest` (without grouping by package or dir), we are already covering a very important gap in our tooling.

What do you think about these? thanks!

**Comment by moyodiallo** [comment]
>First, the UI. It seems that it's a CLI flag to add to build and that will display the missing external libs. How is it intended to be used: just dune build --external-lib-deps? We're already displaying an error message when an external library is not found. What is the expected difference with this flag?

Yep this is an issue, because the command is not expected to fail.

**Comment by moyodiallo** [comment]
>I think I would have expected something a separate command, or a new output in dune describe. I understand that this is less precise: we can't find dependencies of a single target and we can't see behind (select) but is it important to include external libraries that are not going to be selected by the build? But we can probably filter by package by considering only the relevant rules, which seems enough to start.

Yes, It's better to have a separate command, like @rgrinberg propose a good starting point is `dune describe`and tweak `Lib.t` to get `Kind_db.t`. I don't think `select` is relevant because I went to `dune build` because of the idea capturing the external libs during the resolve.

**Comment by moyodiallo** [comment]
>I don't understand the dir argument either. what does this enable? can we start without this? it seems to me that if we first try to collect all external lib dependencies referred to by @install or @runtest (without grouping by package or dir), we are already covering a very important gap in our tooling.

This argument was put to group by package for [opam-dune-lint](https://github.com/ocurrent/opam-dune-lint). Starting with `dune describe`, It won't be a big deal because we have all information needed.

**Comment by emillon** [comment]
Sure, it's fine to add something to the interface.

**Comment by moyodiallo** [comment]
@emillon you were right about `Select`, it make sense to know the libraries behind it. It's included in the last changes.
Feel free about factorizing the code.

File: src/dune_rules/lib.mli [hunk 1]
```diff
@@ -41,6 +41,13 @@ val hash : t -> int
 
 val project : t -> Dune_project.t option
 
+module Kind_db : sig
+  type t =
+    | Installed_libs
+    | Public_libs
+    | Project_libs
```
Comment (emillon) [inline src/dune_rules/lib.mli:48]: is the distinction between project and public libs used? I wonder if we could do without this `kind` field. we can check if a db is the installed_lib ones by checking if its parent is None maybe?

File: bin/external_lib_deps.ml [hunk 2]
```diff
@@ -1,28 +1,217 @@
...
+                         | Optional -> None
+                         | Required -> Some (Lib_name.package_name name))
+                  |> List.map ~f:Package.Name.to_string
+                  |> String.concat ~sep:" ")
+             ]))
+  else if sexp then (
+    if not by_dir then User_error.raise [ Pp.textf "--sexp requires --by-dir" ];
+    let sexp =
+      external_resolved_libs
+      |> List.map ~f:(fun (dir, libs) ->
```
Comment (emillon) [inline bin/external_lib_deps.ml:155]: This is duplicated with the branch below. I would recommend defining an appropriate data structure, build it in both cases, and print it differently (using to_dyn in one case)

**Review by emillon — COMMENTED** [review]
thanks. I think that it's going in the right direction. In addition to the comments I have 2 general remarks:
- I wonder if this should be a `describe` target rather than a top-level command.
- please add tests for this feature, including for corner cases like "optional" and CLI errors (if relevant)

File: bin/external_lib_deps.ml [hunk 3]
```diff
@@ -1,28 +1,217 @@
...
+      |> libs_dir_to_map
+      |> Dir_map.filter ~f:(fun m -> not (Lib_name.Map.is_empty m))
+      |> Dir_map.to_dyn (fun libs -> Lib_name.Map.to_dyn Kind.to_dyn libs)
+      |> Sexp.of_dyn
+    in
+    Format.printf "%a@." Pp.to_fmt
+      (Sexp.pp (Sexp.List [ Sexp.Atom context_name; sexp ])))
+  else (
+    if by_dir then
+      User_error.raise [ Pp.textf "--by-dir cannot be used without --sexp" ];
```
[the remark this answers is not in the document]
Comment (moyodiallo) [reply bin/external_lib_deps.ml:171]: > I wonder if this should be a `describe` target rather than a top-level command.

I don't get what you mean.

**Comment by moyodiallo** [comment]
>  wonder if this should be a describe target rather than a top-level command.

I don't get what you mean.

File: bin/external_lib_deps.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (moyodiallo) [reply bin/external_lib_deps.ml:171]: I was trying keep the old interface, there's no by-dir plain output. I asked myself if the `by-dir` is necessary because only `sexp` could be ok, I think.

**Comment by moyodiallo** [comment]
The sub-command `external-lib-deps` is going to move in `describe`. Any suggestion about the spec ?

File: src/dune_lang/action.ml [hunk 4]
```diff
@@ -398,6 +398,31 @@ let rec map_string_with_vars t ~f =
   | Pipe (o, ts) -> Pipe (o, List.map ts ~f:(map_string_with_vars ~f))
   | Cram sw -> Cram (f sw)
 
+let rec list_of_string_with_vars t =
```
Comment (emillon) [inline src/dune_lang/action.ml:401]: looks like there's a bunch of unrelated code that has been pulled in?

Comment (moyodiallo) [reply src/dune_lang/action.ml:401]: Recycling the old tests, I find out there was `lib-available:<library-name>` and it's considered like optional dependency.

Comment (emillon) [reply src/dune_lang/action.ml:401]: When you cherry-picked old tests in a8b62720b907d23a8373ea30b261cbded42ed318, the commit also added some changes in `src/`. Can you remove those?

Comment (moyodiallo) [reply src/dune_lang/action.ml:401]: I didn't a cherry-pick, the tests was not consequent. So I just did a copy paste mentioning it was the all tests.

Comment (emillon) [reply src/dune_lang/action.ml:401]: OK I understand now. I don't think that this approach is going to work, since it only collects pforms in `(rule)`. Can we stick to the simpler version (that does not try to collect lib names in pforms) for 3.5? Otherwise we can try something that somehow folds over pforms in all stanzas, but this is going to be more involved so we can't do that before 3.6.

File: bin/describe.ml [hunk 5]
```diff
@@ -697,6 +890,16 @@ module What = struct
       , "builds a given FILE and prints the preprocessed output"
       , let+ s = filename in
         Pp s )
+    ; ( "external-lib-deps"
+      , []
+      , "Print out external libraries needed to build the project. It's an \
+         approximated set of libraries."
+      , return External_lib_deps )
+    ; ( "missing-external-lib-deps"
```
Comment (emillon) [inline bin/describe.ml:898]: does `dune-opam-lint` require this second one? we said we'd start with the bare minimum to support that command for now.

Comment (moyodiallo) [reply bin/describe.ml:898]: No. I'm going to retract that.

Comment (moyodiallo) [reply bin/describe.ml:898]: Yeah, even with only external libs we can find missing one without `dune`.

File: src/dune_rules/lib.ml [hunk 6]
```diff
@@ -775,7 +775,7 @@ module rec Resolve_names : sig
   val resolve_dep :
     db -> Loc.t * Lib_name.t -> private_deps:private_deps -> lib Resolve.Memo.t
 
-  val resolve_name : db -> Lib_name.t -> Status.t Memo.t
+  val resolve_name : db -> Lib_name.t -> (Status.t * db) Memo.t
```
Comment (rgrinberg) [inline src/dune_rules/lib.ml:778]: Can you explain why we need this change?

File: src/dune_lang/string_with_vars.mli [hunk 7]
```diff
@@ -10,6 +10,8 @@ open Dune_util
 (** A sequence of text and variables. *)
 type t
 
+val pforms : t -> Pform.t list
```
Comment (rgrinberg) [inline src/dune_lang/string_with_vars.mli:13]: Where is this function being used?

File: src/dune_rules/expander.mli [hunk 8]
```diff
@@ -141,3 +141,5 @@ val expand_locks :
   base:[ `Of_expander | `This of Path.t ] -> t -> Locks.t -> Path.t list Memo.t
 
 val sites : t -> Sites.t
+
+val bindings : t -> value Pform.Map.t
```
Comment (rgrinberg) [inline src/dune_rules/expander.mli:145]: Where is this function being used?

**Comment by emillon** [comment]
It looks like there's still a bit of work for this feature to be stable so I'm marking this for 3.6 to avoid delaying the 3.5 release.

File: src/dune_lang/string_with_vars.mli (hunk 7 above)
Comment (moyodiallo) [reply src/dune_lang/string_with_vars.mli:13]: Yeah Previous commit, I did forgot. I'll remove it.
Thanks.

File: src/dune_rules/expander.mli (hunk 8 above)
Comment (moyodiallo) [reply src/dune_rules/expander.mli:145]: Same thing. Thanks.

File: src/dune_rules/lib.ml (hunk 6 above)
Comment (moyodiallo) [reply src/dune_rules/lib.ml:778]: This is the last `db` where the `lib_name` where searched(or resolved) and with that, we can determine if the `lib_name` is external or not(external if`db.parent` is None).

Yeah I see your concern. I think it would it be better to expand Status.t by having `External` in it and remove `db`.

**Comment by moyodiallo** [comment]
> It looks like there's still a bit of work for this feature to be stable so I'm marking this for 3.6 to avoid delaying the 3.5 release.

Just one more little thing, It's gonna be possible. :grin:

**Comment by moyodiallo** [comment]
I did another review to delete all the parasite code that I introduce(git diff main). And I think it's good for me, unless you have another remarks or concern.

File: src/dune_rules/lib.mli [hunk 9]
```diff
@@ -131,6 +131,8 @@ module DB : sig
 
   val find_even_when_hidden : t -> Lib_name.t -> lib option Memo.t
 
+  val is_external : t -> Lib_name.t -> bool Memo.t
```
Comment (rgrinberg) [inline src/dune_rules/lib.mli:134]: Instead of introducing this function. How about just doing a resolve first (with find_even_when_hidden). Once the library is resolved, you can check if it's external by looking at the status.

**Review by rgrinberg — CHANGES_REQUESTED** [review]

File: src/dune_rules/lib.mli (hunk 9 above)
Comment (moyodiallo) [reply src/dune_rules/lib.mli:134]: `Status.t` don't give the information if the lib is external or not. And expanding status it's not a good idea, I tried it.
Resolve then a check, could end up with the case when the library is not installed(not found). We want to print the information whether or not the library is installed.

Comment (rgrinberg) [reply src/dune_rules/lib.mli:134]: The constructors `Installed` and `Installed_private` correspond to external libraries.

Why is expanding the status not a good idea?

> Resolve then a check, could end up with the case when the library is not installed(not found). We want to print the information whether or not the library is installed.

I don't understand. Is there a test case that corresponds to this situation?

Comment (moyodiallo) [reply src/dune_rules/lib.mli:134]: I said not a good Idea because I tried to have a constructor `External` in the status. It was conflicting whether  the lib is `Found` or `Not_found`.

Yeah I see you suggestion, we doesn't need to expand the status if it's true:
Is it that, if find_even_hidden give `None` the lib is external and if we get a lib, we test whether it's `Installed`?

Comment (moyodiallo) [reply src/dune_rules/lib.mli:134]: If a lib is not found, we can assume it is an external lib.

Comment (rgrinberg) [reply src/dune_rules/lib.mli:134]: > Is it that, if find_even_hidden give None the lib is external and if we get a lib, we test whether it's Installed?

Yes, we can do that.

> If a lib is not found, we can assume it is an external lib.

That's the best we can do.

In general, we don't want to modify any existing code in `lib.ml`. It's pretty complicated already and we'd prefer changes that simplify things or at least add things independently. That's why I'm pushing against these changes.

Comment (moyodiallo) [reply src/dune_rules/lib.mli:134]: I understand, that sounds good for me and it's equivalent to what I was already doing.

**Comment by moyodiallo** [comment]
The irony(git diff main), using `find_even_with_hidden` simplifies a lot.

**Comment by rgrinberg** [comment]
Thanks. I merged after a few cleanups.

**Comment by moyodiallo** [comment]
Thanks All of you. I learnt a lot during this PR.
